crates/turbo-tasks-fs/src/rope.rs

```diff
@@ -532,31 +538,79 @@ impl DeterministicHash for RopeElem {
     }
 }
 
-/// Implements the Read/AsyncRead/Stream/Iterator trait over a Rope.
-#[derive(Debug, Default)]
-pub struct RopeReader {
+/// Implements the [Read]/[AsyncRead]/[Iterator] trait over a [Rope].
+#[derive(Debug)]
+pub struct RopeReader<'a> {
     /// The Rope's tree is kept as a cloned stack, allowing us to accomplish
     /// incremental yielding.
-    stack: Vec<StackElem>,
+    stack: Vec<StackElem<'a>>,
+
+    /// The held root allows us to invert the ownership of Bytes. By storing a
+    /// copy of Rope's root, we can ensure that our borrowed StackElems will
+    /// live at least as long as this RopeReader.
+    held_root: Option<InnerRope>,
 }
 
 /// A StackElem holds the current index into either a Bytes or a shared Rope.
 /// When the index reaches the end of the associated data, it is removed and we
 /// continue onto the next item in the stack.
 #[derive(Debug)]
-enum StackElem {
-    Local(Bytes),
-    Shared(InnerRope, usize),
-}
-
-impl RopeReader {
-    fn new(inner: &InnerRope, index: usize) -> Self {
-        if index >= inner.len() {
-            Default::default()
+enum StackElem<'a> {
+    Local(&'a Bytes, usize),
+    Shared(&'a InnerRope, usize),
+}
+
+impl<'a> RopeReader<'a> {
+    /// Constructs a new RopeReader, taking ownership of the Rope's internal
+    /// memory and freeing us from the lifetime constraints of the Rope.
+    fn new(rope: &Rope) -> RopeReader<'static> {
+        // SAFETY: We transmute the InnerRope into a 'static lifetime so that it is
+        // usable without keeping a reference to the Rope alive. This is safe because
+        // the RopeReader itself holds an Arc clone of the data, ensuring the data will
+        // live at least as long as the reader.
+        let borrow = unsafe { std::mem::transmute::<&InnerRope, &'static InnerRope>(&rope.data) };
```

Comment:
This is not safe since you are holding a reference to the passed `InnerRope` and not to the cloned `InnerRope` that's store in `held_root`. You would need to move the `reader.held_root = Some(borrow.clone());` up and borrow from that instead. But that doesn't work for two reasons:
* `new_borrow` returns a new object.
* It would require you to return `Pin<Box<RopeReader<...>>>` since you need to prevent the memory location of `held_root` from moving.

It's probably better to hold references to the content of the `Arc` instead (`&[RopeElem]`) since they won't move as long as you are holding the Arc

Reply:
But in the end I would avoid the unsafe logic here at all by adding a `StackElem::Owned(usize)` which looks up the index in `held_root` (probably rename it)

Reply:
> But in the end I would avoid the unsafe logic here at all by adding a StackElem::Owned(usize) which looks up the index in held_root (probably rename it)

I think this'll still require a `transmute`, so that I can cast references from `&self.held_root[index]` into the static lifetime, too.